### PR TITLE
close #1031 api for pie_chart crew event dashboard

### LINKED ISSUE #1031: API for pie_chart crew event dashboard

### PR DESCRIPTION
**Pull request by LengTech11**
##### GET: {{BASE_URL}}/api/v2/operator/taxons/115/pie_chart_event_aggregators?chart_type=participation

```json
{
    "data": {
        "id": "115",
        "type": "pie_chart_event_aggregator",
        "attributes": {
            "id": "115",
            "type": "participation",
            "value": [
                {
                    "product_id": 128,
                    "checked_in_ticket": 1,
                    "total_ticket": 4
                },
                {
                    "product_id": 132,
                    "checked_in_ticket": 2,
                    "total_ticket": 2
                }
            ]
        }
    }
}
```

##### GET: {{BASE_URL}}/api/v2/operator/taxons/115/pie_chart_event_aggregators?chart_type=gender

```json
{
    "data": {
        "id": "115",
        "type": "pie_chart_event_aggregator",
        "attributes": {
            "id": "115",
            "type": "gender",
            "value": [
                {
                    "genders": "female",
                    "gender_count": 2
                },
                {
                    "genders": "male",
                    "gender_count": 1
                }
            ]
        }
    }
}
```

##### GET: {{BASE_URL}}/api/v2/operator/taxons/115/pie_chart_event_aggregators?chart_type=entry_type

```json
{
    "data": {
        "id": "115",
        "type": "pie_chart_event_aggregator",
        "attributes": {
            "id": "115",
            "type": "entry_type",
            "value": [
                {
                    "product_id": 128,
                    "entry_types": "normal",
                    "check_in_count": 1
                },
                {
                    "product_id": 132,
                    "entry_types": "normal",
                    "check_in_count": 1
                },
                {
                    "product_id": 132,
                    "entry_types": "VIP",
                    "check_in_count": 1
                }
            ]
        }
    }
}
```

### CODE REVIEW

File: app/interactors/spree_cm_commissioner/operator_dashboard_interactor.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module SpreeCmCommissioner
+  class OperatorDashboardInteractor < BaseInteractor
+    def self.call
+      result = OperatorDashboardDataQuery.call
+
+      result.group_by { |row| row['chart_type'] }.map do |chart_type, rows|
+        create_pie_chart_for(chart_type: chart_type, rows: rows)
+      end
+    end
```
Comment (theachoem) [inline app/interactors/spree_cm_commissioner/operator_dashboard_interactor.rb:9]:
```rb
def call
   dashboard_query_result = dashboard_query

    context.pie_charts = result.group_by { |row| row['chart_type'] }.map do |chart_type, rows|
        create_pie_chart_for(chart_type: chart_type, rows: rows)
     end
end

def dashboard_query
   OperatorDashboardDataQuery.call
end
```

```rb
allow_any_instance_of(OperatorDashboardInteractor).to_recieve(:dashboard_query).and_return([
  {'chart_type': 'gender', pies: []}, {'chart_type': 'gender', pies: []}, 
])

context = OperatorDashboardDataQuery.call
expect(context.pie_charts[0].chart_type).to eq 'gender'
```

File: app/controllers/spree/api/v2/operator/operator_pie_charts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  module Api
+    module V2
+      module Operator
+        class OperatorPieChartsController < ::Spree::Api::V2::ResourceController
+          before_action :require_spree_current_user, only: %i[index]
+
+          def index
+            objects = SpreeCmCommissioner::OperatorPieChartInteractor.call
+            # render_serialized_payload(200) { serialize_resource(objects) }
+            render_serialized_payload(201) do
```
Comment (channainfo) [inline app/controllers/spree/api/v2/operator/operator_pie_charts_controller.rb:11]: Use 200 status code

File: app/interactors/spree_cm_commissioner/operator_pie_chart_interactor.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+module SpreeCmCommissioner
+  class OperatorPieChartInteractor < BaseInteractor
+    def self.call
```
Comment (channainfo) [inline app/interactors/spree_cm_commissioner/operator_pie_chart_interactor.rb:3]: Use def call instead.

File: app/controllers/spree/api/v2/operator/operator_pie_charts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+module Spree
+  module Api
+    module V2
+      module Operator
+        class OperatorPieChartsController < ::Spree::Api::V2::ResourceController
+          before_action :require_spree_current_user, only: %i[index]
+
+          def index
+            objects = SpreeCmCommissioner::OperatorPieChartInteractor.call
```
Comment (channainfo) [inline app/controllers/spree/api/v2/operator/operator_pie_charts_controller.rb:9]: This method should receive an event , a date and a chart type( gender, profession,...)

File: app/queries/spree_cm_commissioner/operator_pie_chart_data_queries.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    def self.list_all_guests
+      'SELECT * FROM cm_guests WHERE line_item_id IS NOT NULL'
+    end
+
+    def self.check_in_data
+      'SELECT * FROM cm_check_ins'
+    end
+
+    def self.dashboard_data
+      "SELECT list_all_guest.*, check_in_data.guest_id AS has_check_ins,
```
Comment (channainfo) [inline app/queries/spree_cm_commissioner/operator_pie_chart_data_queries.rb:12]: Use the heredoc syntax for multi line strings

File: app/queries/spree_cm_commissioner/operator_pie_chart_data_queries.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+         WHEN dashboard_data.entry_type = 0 THEN 'normal'
+         WHEN dashboard_data.entry_type = 1 THEN 'VIP'
+         ELSE NULL
+       END AS entry_types
+       FROM dashboard_data
+       GROUP BY dashboard_data.entry_type"
+    end
+
+    def self.pie_charts_gender
+      "SELECT genders.genders, COUNT(*) as pie
```
Comment (channainfo) [inline app/queries/spree_cm_commissioner/operator_pie_chart_data_queries.rb:42]: Find a way to reuse this code by passing a type in.

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/queries/spree_cm_commissioner/operator_pie_chart_data_queries.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
+module SpreeCmCommissioner
+  module OperatorPieChartDataQueries
+    def self.operator_pie_charts_queries
```
Comment (channainfo) [inline app/queries/spree_cm_commissioner/operator_pie_chart_data_queries.rb:3]: let's not calculate all in a single method. we should have to pass in a chart type for example a chart for the gender. we return only that specific chart. The app needs to send to the API with a corresponding chart type.

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/interactors/spree_cm_commissioner/operator_pie_chart_interactor.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    def pie_chart_query
+      OperatorPieChartDataQuery.call
+    end
+
+    def create_pie_charts(data)
+      data.group_by { |row| row['chart_type'] }.map do |chart_type, chart_data|
+        create_pie_chart_for(chart_type: chart_type, rows: chart_data)
+      end
+    end
+
```
Comment (channainfo) [inline app/interactors/spree_cm_commissioner/operator_pie_chart_interactor.rb:17]:
```
# chart_type: gender | entry_type | 
 def create_pie_chart_for(chart_type:, rows:)
```

Comment (channainfo) [reply app/interactors/spree_cm_commissioner/operator_pie_chart_interactor.rb:17]:
```
Rails.cache.fetch("#{chart_type}-#{event_id}", expires_in: 30.minutes) do
   create_pie_chart_for(chart_type: ...)
end

```

**Review by channainfo — CHANGES_REQUESTED** [review]

**Pull request merged**